### PR TITLE
internal/provisioner: support using ContourDeployment.Config

### PR DESCRIPTION
**Pull request by skriss**
When a GatewayClass has a parametersRef
to a ContourDeployment resource that has
Config specified, use that Config when
provisioning Gateways and their associated
ContourConfigurations.

Signed-off-by: Steve Kriss <krisss@vmware.com>

### CODE REVIEW

File: internal/provisioner/controller/gateway.go [hunk 1]
```diff
@@ -323,3 +332,22 @@ func (r *gatewayReconciler) ensureContourDeleted(ctx context.Context, contour *m
 
 	return errs
 }
+
+func (r *gatewayReconciler) getGatewayClassParams(ctx context.Context, gatewayClass *gatewayapi_v1alpha2.GatewayClass) (*contour_api_v1alpha1.ContourDeployment, error) {
+	if !isContourDeploymentRef(gatewayClass.Spec.ParametersRef) {
```
Comment (sunjayBhatia) [inline internal/provisioner/controller/gateway.go:337]: is this case possible? since GatewayClasses won't be accepted if they have an invalid parametersref and we won't reconcile a Gateway for a GatewayClass that has not been accepted?

Comment (skriss) [reply internal/provisioner/controller/gateway.go:337]: Yeah probably not, I think this function call is effectively just handling nil-checking on the parametersref right now. Can simplify to just do that directly.

Comment (skriss) [reply internal/provisioner/controller/gateway.go:337]: Thinking about it some more, there probably is a (admittedly pretty hard-to-trigger) race condition here, if a GatewayClass has been updated from a valid ContourDeployment ref to a ref to some other resource type but hasn't yet reconciled / marked "Accepted: false", and ~simultaneously a Gateway reconcile is triggered that uses that GatewayClass. Based on that I'm thinking it's safer to keep the check here (it's just checking fields, no cache/API calls so not expensive to execute), but I'll add some comments.

Comment (sunjayBhatia) [reply internal/provisioner/controller/gateway.go:337]: ah yep that is a good point about the race, yeah worth adding some comments (and maybe a test if possible) as to what is expected in this edge case

**Review by sunjayBhatia — APPROVED** [review]
